cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py

```diff
@@ -257,11 +258,27 @@ def update_kubelet_status():
     ''' There are different states that the kubelet can be in, where we are
     waiting for dns, waiting for cluster turnup, or ready to serve
     applications.'''
-    if (_systemctl_is_active('snap.kubelet.daemon')):
+    services = [
+        'kubelet',
+        'kube-proxy'
+    ]
+    for service in services:
+        daemon = 'snap.{}.daemon'.format(service)
+        if not _systemctl_is_active(daemon):
+            hookenv.log("Service {} id down. Starting it.".format(daemon))
+            sleep(10)
```

Comment:
Your comment prompted me to look more into this. In short, yes a waiting period is needed but we should not try to restart the services like this here.

In more detail, the sleep creates a waiting period in which we are waiting for a potential failure. To see why this sleep is needed one can remove it completely, introduce a wrong variable in /var/snap/<my-service>/current/args, systemctl stop <my-service>, and then trigget a hooks/update-status. The update status will see the service being down and will restart it. However, <my-service> will not fail immediately so the status message will report that everything is OK. On the next status-update cycle we will have the same, service will restart but will not fail in time and we will be reporting that everything is ok. So, the sleep creates a wait period in which we expect the service to fail. This is problematic because there is no proper way to estimate what this waiting period would be. 10 seconds is enough for a wrong argument to be detected on an aws instance but overall there is no safe guess.

My suggestion is to trust systemd to restart failing services. If a service is down that means that systemd is failing to restart the service, in which case the admin should take a look. We could assist the admin by adding some actions to inspect the services and restart them.